### DISCUSSION TITLE
Sample code for unit testing adapters please?

### DISCUSSION THREAD

**Discussion by jeffvictorino** [state-adapt/state-adapt#64] [General]
For the past 2 weeks I've been implementing a new feature in our angular application. Using state adapt 2.0. 

I'm now able to implement a basic store and adapter, and trying to write unit tests in jasmine. 

With a basic adapter like
`export const productAdapter = createAdapter<Product[]>()({
  changeQuantity: (state, { name, quantity }: QuantityChange) =>
    state.map(product => (product.name === name ? { ...product, quantity } : product)),
  addProduct: (state, product: Product) => [...state, product],
  removeProduct: (state, { name }: Product) =>
    state.filter(product => product.name !== name),
  selectors: {
    totalPrice: products =>
      products.reduce((total, { price, quantity }) => total + price * quantity, 0),
  },`
  
  How would I go and test the functions `changeQuantity`, `addProduct`, and the selectors in a unit test? 
  
  Thanks a lot in advance.

**Comment by mfp22** [comment]
Adapters are just objects containing pure functions, so you can just test them directly.

My tests might not be the most amazing or readable, but here are some from this repo itself.

(you can ignore the type testing on this one) 
https://github.com/state-adapt/state-adapt/blob/main/libs/core/src/lib/adapters/create-adapter.function.spec.ts

(you can ignore the run time counts on this one)
https://github.com/state-adapt/state-adapt/blob/main/libs/core/src/lib/adapters/build-adapter.function.spec.ts

https://github.com/state-adapt/state-adapt/blob/main/libs/core/src/lib/adapters/join-adapters.function.spec.ts

https://github.com/state-adapt/state-adapt/blob/main/libs/rxjs/src/lib/global-store/state-adapt.spec.ts

You can ignore the selectors cache object I'm passing in. All it does is make the selectors memoize. It doesn't affect the results.

**Comment by jeffvictorino** [comment]
Thanks Mike,  I realised wallaby.js is acting up and just not picking my tests up properly.

You are right, adapters just contain pure functions and it's easy to state by passing state and other parameters.
